### DISCUSSION TITLE
Decompose loyalty program setup metadata

### DISCUSSION THREAD

**Discussion by mnunezdm** [forcedotcom/cli#2960] [Ideas]
[Loyalty Program Setup](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_loyaltyprogramsetup.htm) the main configuration metadata type for loyalty does not allow decomposition when retrieving deploying metadata

As it can be tough to manage when the setup gets very complex I think it would be beneficial to allow the decomposing/recomposing of this metadata to maintain legibility and better source control

I have tried doing a simple script to manually decompose/recompose the metadata but this adds complexity to the development and deployment process

**Comment by mcarvin8** [comment]
@mnunezdm - Depending on Salesforce's response, I imagine they would start with a Beta decomposition service before it making it GA if they proceed with this. 

If you want to decompose more immediately and would like to try a 3rd party CLI plugin, you can give my plugin a stab assuming you are using a Salesforce DX Git repo - https://github.com/mcarvin8/sf-decomposer

If you create the JSON file in your repo root for the Hook, it can automatically decompose and recompose when running the retrieve and deploy CLI commands. Thereby, this shouldn't add any complexity to your deployment scripts outside of installing the plugin beforehand and configuring the JSON file for the hook. 

The decomposed files will be named with SHA hashes. Based on the Metadata API guide, I could release a minor plugin update to add the unique ID element for [LoyaltyProgramProcess[]](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_loyaltyprogramsetup.htm#subtype_LoyaltyProgramProcess), which looks to be `processName`.

The meta suffix for this MD would be `loyaltyProgramSetup`.

**Comment by mcarvin8** [comment]
If you'd like, you can try sf-decomposer@4.1.8 , which I just released today to add support for the loyalty program setup unique ID element. I created it using the example meta file provided in the metadata api guide for this type.

**Reply by mnunezdm** [reply]
@mcarvin8  i'll give it a try! just one thing, are you just decomposing into loyalty program process? or you are also decomposing parameters and rules

<img width="720" alt="image" src="https://github.com/user-attachments/assets/9f5b0f45-bb3d-42a6-8f13-55d95e728190">

This is what my "decomposition" looks like

**Reply by mcarvin8** [reply]
At the moment, my decomposition is just going into the 1st level, which in this case would be into loyalty program process.

I've been thinking about further decomposition and what would be the best way to program that for a more general solution.

My Salesforce decomposer depends on another NodeJS package I developed which is a [general XML disassembler](https://github.com/mcarvin8/xml-disassembler). I'll take a look into what updates I would need to do to decompose further past the 1st level and recompose successfully. If you have any suggestions, let me know on the xml-disassembler package. 

I'm thinking I'll either need to update the nested element function to continue until there are no nested elements anymore or recursively call my disassembler until no more sub-directories are created. I have a "--postpurge" flag to delete XMLs after decomposing them so I'll have to use that.

**Reply by mcarvin8** [reply]
@mnunezdm I think I was finally able to crack how to decompose loyalty program setup by program processes and rules similar to how you are doing it. 

I just released my sf-decomposer@5.10.0 to make this the default strategy for loyaltyProgramSetup.

Main thing to note is to successfully recompose the original XML, I have to delete the decomposed files during the process. I usually let users set a "--postpurge" flag to let them choose but multi-level decomposition and recompostion only worked when I delete the files after I'm done with them. This can easily be discarded in a version control system or in a CI pipeline. So I would recommend only recomposing in a CI pipeline or git repo where you can discard the deleted decomposed files. 

If you want to give this a shot, let me know if you have issues with your loyalty program setup decomposing or recomposing correctly to deploy.  

Green = decomposed files
Yellow = original file

![image](https://github.com/user-attachments/assets/217070c8-6c1d-44a4-afaa-4ab215238534)

**Reply by mnunezdm** [reply]
We are going to give them a try, I'm assuming that you need to manually launch the commands and cannot be launched in a `pre deploy` / `post retrieve` hook

**Reply by mcarvin8** [reply]
I have hooks configured if you create a ".sfdecomposer.config.json" in the root of your project like such:

```json

{
  "metadataSuffixes": "loyaltyProgramSetup",
  "decomposedFormat": "xml"
}
```

This will decompose automatically after "sf project retrieve" and recompose automatically before "sf project deploy". 

See the README in https://github.com/mcarvin8/sf-decomposer

Please see the note regarding adding decomposed files to your ".forceignore" file. This is required to avoid errors when the hooks fire during CLI commands.

**Comment by VivekMChawla** [comment]
@mnunezdm - I've added this to our New Source Behavior backlog for consideration. [W-16511226](https://gus.lightning.force.com/lightning/r/a07EE00001z8t6CYAQ/view).

If we proceed, a source decomposition design will first be presented to the community for feedback. 

I'll post another message to this discussion if/when that happens.

Thank you!
